File: apps/frontend/src/app/foundation/payments/payment.service.ts
import { Injectable, inject } from '@angular/core';
import { Observable } from 'rxjs';

import { ApiClient } from '@foundation/networking';
import {
  CreatePaymentIntentRequest,
  PaymentIntent,
  PaymentMethod,
  PaymentProviderResult,
} from './payment.models';

@Injectable({ providedIn: 'root' })
export class PaymentService {
  private readonly api = inject(ApiClient);

  listPaymentMethods(): Observable<PaymentMethod[]> {
    return this.api.get<PaymentMethod[]>('/payments/methods');
  }

  createIntent(payload: CreatePaymentIntentRequest): Observable<PaymentIntent> {
    return this.api.post<PaymentIntent, CreatePaymentIntentRequest>('/payments/intents', payload);
  }

  confirmIntent(paymentIntentId: string): Observable<PaymentProviderResult> {
    return this.api.post<PaymentProviderResult>(
      `/payments/intents/${encodeURIComponent(paymentIntentId)}/confirm`,
      {},
    );
  }
}

File: apps/ios/SpotLink/Sources/SpotLink/Features/Payments/PaymentService.swift
import Foundation

// MARK: - Payment Service

public final class PaymentService: Sendable {
    private let apiClient: APIClientProtocol

    public init(apiClient: APIClientProtocol) {
        self.apiClient = apiClient
    }

    public func capabilities() async throws -> PaymentCapabilities {
        try await apiClient.get("/payments/capabilities", query: nil)
    }

    public func listPaymentMethods() async throws -> [PaymentMethod] {
        try await apiClient.get("/payments/methods", query: nil)
    }

    public func createIntent(_ request: CreatePaymentIntentRequest) async throws -> PaymentIntent {
        try await apiClient.post("/payments/intents", body: request)
    }

    public func confirmIntent(_ intentId: String) async throws -> PaymentProviderResult {
        try await apiClient.post("/payments/intents/\(intentId)/confirm", body: EmptyPayload())
    }

    public func cancelIntent(_ intentId: String) async throws -> PaymentProviderResult {
        try await apiClient.post("/payments/intents/\(intentId)/cancel", body: EmptyPayload())
    }
}

private struct EmptyPayload: Encodable {}
